Proposed changes for version 2.2.0 of the data models.

geoffmartin-adlink opened the pull request:
* Incorporates changes to address issues 6, 7, 12 and 14 as well as changes for consistency.
* Version increased to 2.2.0 as 2.1.0 may have already been deployed by Alex.
* Some of Alex's proposed changes have been included in this PR. Others need further discussion.

lee-hemingway-adlink commented on definitions/TagGroup/com.adlinktech.vision.system/2.2.0/EngineConfigTagGroup.json:
```diff
@@ -0,0 +1,187 @@
+            }
+        ]
+    },
+    {"typedefinition":"EngineStreamConfig",
```
This typedefinition is missing some new lines.

lee-hemingway-adlink commented on definitions/TagGroup/com.adlinktech.vision.system/2.2.0/ConfigActualStateTagGroup.json:
```diff
@@ -0,0 +1,35 @@
+        "tags":[
+            {
+                "name":"configuration_data",
+                "description":"Detection Box Data (the results)",
```
Don't think we use this but this description looks like a c&p error.

lee-hemingway-adlink commented on definitions/TagGroup/com.adlinktech.vision.system/2.2.0/ModelConfigRequestTagGroup.json:
```diff
@@ -0,0 +1,48 @@
+       {
+          "name": "hardware",
+          "description": "The hardware to run the model on: CPU, GPU, VPU or AUTO (optional -- defaults to AUTO)",
+	  "optional":true,
```
Indentation issue.

lee-hemingway-adlink commented on definitions/TagGroup/com.adlinktech.vision.system/2.2.0/ModelConfigRequestTagGroup.json:
```diff
@@ -0,0 +1,48 @@
+       {
+          "name": "threshold",
+          "description": "The threshold at which model outputs should be published (optional -- defaults to 0.50)",
+	  "optional":true,
```
Indentation issue.

lee-hemingway-adlink commented on definitions/TagGroup/com.adlinktech.vision.system/2.2.0/ModelConfigRequestTagGroup.json:
```diff
@@ -0,0 +1,48 @@
+       {
+          "name": "reload",
+          "description": "A boolean indicating whether or not the model should be dynamically loaded (optional -- defaults to false)",
+	  "optional":true,
```
Indentation issue.

lee-hemingway-adlink commented on definitions/TagGroup/com.adlinktech.vision.system/2.2.0/ConfigActualStateTagGroup.json:
```diff
@@ -0,0 +1,35 @@
+            {
+                "name":"configuration_data",
+                "description":"Detection Box Data (the results)",
+                "kind":"NVP_SEQ",
```
Again, we don't use this but this is possibly supposed to be MULTI_DIM_NVP.

lee-hemingway-adlink commented on definitions/TagGroup/com.adlinktech.vision.system/2.2.0/AlarmTagGroup.json:
```diff
@@ -0,0 +1,40 @@
+        {
+            "name":"meta",
+            "description":"Additional metadata",
+	    "optional":true,
```
Indentation issue.

lee-hemingway-adlink commented on definitions/TagGroup/com.adlinktech.vision.system/2.2.0/DeviceInfoTagGroup.json:
```diff
@@ -0,0 +1,87 @@
+            "unit":"DeviceKind"
+        },
+        {
+            "name":"protocol",
```
The PR says that issue 8 has been addressed but I don't see possible enum values defined/documented anywhere for this and other enums.

geoffmartin-adlink replied:
Issue 8 hasn't been addressed. It's a documentation issue and it appears Ronald hasn't provided the means to regenerate the documentation.

aas5670 commented on definitions/TagGroup/com.adlinktech.vision.inference/2.2.0/DetectionPointTagGroup.json:
```diff
@@ -0,0 +1,186 @@
+        "context":"com.adlinktech.vision.inference",
+        "qosProfile":"telemetry",
+        "version":"2.2.0",
+        "description":"Inference engine results for object detection model outputing points. The stream_id is not part of this TagGroup and assumed to be part of the run-time flowId.",
```
'outputing' -> 'outputting'

geoffmartin-adlink commented:
I've also (re)added a DataCodes tag group under com.adlinktech.vision.inference for Smart Pallet.

geoffmartin-adlink commented:
I think I've now incorporated all of the requested changes from @hendrenja into this PR. Let me know if I've missed anything. Due to the change in folders making things difficult to compare (I've setup this branch so we can work from it while the models are still under discussion) I've attached a patch file showing the differences between the actual files.

[patch.txt](https://github.com/ADLINK-DEV/ADLINK-VisionDataModels-Reference/files/6050603/patch.txt)

@hendrenja and @hansvanthag please let me know if you're both happy with this or if you'd like anything changing.

hansvanthag commented:
Just to make sure: I see that "BinaryArtefact" has the 'event' QoS whereas "OpenVINOModelArtifact" is of the kind 'telemetry' ..

As both are 'Artefacts' .. it looks somewhat strange. 

Now as telemetry data is (similar to events) reliably delivered both (event and telemetry)  might be appropriate, but if (and only if)  updates come in quickly, only the most recent 'artefact' will survive (due to a 1-place buffer at the receiving side that can be easily overwritten, which is of course fine for typical telemetry where 'previous-values' are less important than the 'most recent' one ..

hansvanthag commented:
and more observation: DetectionBox (and perhaps others, but I only spotted this tagGroup) has a field with a name 'uuid' which is more a description of its type/unit than what this tag is about (as is the case for nearly all other 'uuid-type' tags that DO have a proper name).

Its not wrong (as you're free to name your tags as you want) yet I'd consider this almost 'bad practice'

geoffmartin-adlink commented:
> Just to make sure: I see that "BinaryArtefact" has the 'event' QoS whereas "OpenVINOModelArtifact" is of the kind 'telemetry' ..
> 
> As both are 'Artefacts' .. it looks somewhat strange.
> 
> Now as telemetry data is (similar to events) reliably delivered both (event and telemetry) might be appropriate, but if (and only if) updates come in quickly, only the most recent 'artefact' will survive (due to a 1-place buffer at the receiving side that can be easily overwritten, which is of course fine for typical telemetry where 'previous-values' are less important than the 'most recent' one ..

In this case all I've changed is the version number and context (as its not really inference related) of the tag group. The OpenVINOModelArtifact tag group was created by the UK team and the BinaryArtefact by Alex I belive which may explain the discrepancy.

@hendrenja would using the telemetry profile for the BinaryArtefact tag group work for you?

geoffmartin-adlink commented:
> and more observation: DetectionBox (and perhaps others, but I only spotted this tagGroup) has a field/tag with a name 'uuid' which is more a description of its type/unit than what this tag is about (as is the case for nearly all other 'uuid-type' tags that DO have a proper name i.c.w. a "UUID" unit specification).
> 
> Its not wrong (as you're free to name your tags as you want) yet I'd consider this almost 'bad practice'

These fields were copied from @hendrenja pull request so I'll defer to Alex on what the field should be named.

hendrenja commented:
@geoffmartin-adlink we are no longer using `BinaryArtifact` so it can be discarded or updated to use `telemetry`

lee-hemingway-adlink commented:
> @geoffmartin-adlink we are no longer using `BinaryArtifact` so it can be discarded or updated to use `telemetry`

Our aea-aws-model-streamer still uses it as that app hasn't been updated for a long time. Would suggest keeping and changing to telemetry for now.

geoffmartin-adlink commented:
I've committed an update with the changes discussed on today's call:

* Removed CaptureStart and CaptureStop tag groups.
* Renamed uuid in inference tag groups to inference_id.
* Renamed parent_id in inference tag groups to parent_inference_id.
* Set inference_id and parent_inference_id units to be uuidv4.
* Removed DataCode tag group and added DataCode tag into DetectionBox instead.
* Added Phi tag to PhysicalCoords type definition in DetectionBox tag group.
* Added PhysicalCoords type to DetectionPoint tag group with x, y and z tags.

geoffmartin-adlink commented:
Committed a few extra tweaks:

* Changed the type of frame_id in VideoFrame to match frame_id tags in inference tag groups.
* Change the name of the GraphNode type definition in DetectionPoint tag group to GraphEdge as it represents an edge not a node. Also updated the vertex indexes to be UINT64 as indicies will always be positive.

OctanMan commented on definitions/TagGroup/com.adlinktech.vision.system/2.2.0/TrainingStreamerConfigTagGroup.json:
```diff
@@ -0,0 +1,53 @@
+                "unit": "NA"
+            },
+            {
+                "name": "StreamId",
+                "description": "List of stream id",
```
The name and description don't seem right here. Maybe they should be
```suggestion
                "name": "StreamInformation",
                "description": "Information about the stream",
```

geoffmartin-adlink replied:
Changed. I also changed the description of the overall tag group.

OctanMan commented on definitions/TagGroup/com.adlinktech.vision.capture/2.2.0/CameraConnectRequestTagGroup.json:
```diff
@@ -0,0 +1,72 @@
+        "context":"com.adlinktech.vision.capture",
+        "qosProfile":"event",
+        "version":"2.2.0",
+        "description":"Camera Add Request: sent when a camera should be added to the river",
```
```suggestion
        "description":"Camera Remove Request: sent when a camera should be removed from the Data River",
```

geoffmartin-adlink replied:
I've made this change on the branch.

geoffmartin-adlink commented:
I've made a new commit to this branch to support the integration of the NVIDIA DeepStream Analytics plugin. Changes made include:

* Added new tag groups (RegionOfInterest, Line and LineCrossed).
* Modified how information from a tracker is represented in the DetectionBox tag group to support additional information on tracked objects.

Given the amount and nature of the changes to date I've attached a diff file to show the differences more clearly.

[VisionModels.txt](https://github.com/ADLINK-DEV/ADLINK-VisionDataModels-Reference/files/6117155/VisionModels.txt)

geoffmartin-adlink commented:
Proposed changes has been discussed with Hans and he's happy with them. Merging to master to allow progress to be made in moving our apps to the new data models.
